frontend/test/lit/test_mlir_decomposition.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This file performs the frontend lit tests that the peephole transformations are correctly lowered.
+
+We check the transform jax primitives for each pass is correctly injected
+during tracing, and these transform primitives are correctly lowered to the mlir before
+running -apply-transform-sequence. 
+"""
+
+# RUN: %PYTHON %s | FileCheck %s
+
+# pylint: disable=line-too-long
+
+import platform
+import shutil
+
+import pennylane as qml
+from lit_util_printers import print_jaxpr
+
+from catalyst import qjit
+from catalyst.debug import get_compilation_stage
+from catalyst.device import get_device_capabilities
+from catalyst.utils.runtime_environment import get_lib_path
+
+
+def flush_peephole_opted_mlir_to_iostream(QJIT):
+    """
+    The QJIT compiler does not offer a direct interface to access an intermediate mlir in the pipeline.
+    The `QJIT.mlir` is the mlir before any passes are run, i.e. the "0_<qnode_name>.mlir".
+    Since the QUANTUM_COMPILATION_PASS is located in the middle of the pipeline, we need
+    to retrieve it with keep_intermediate=True and manually access the "2_QuantumCompilationPass.mlir".
+    Then we delete the kept intermediates to avoid pollution of the workspace
+    """
+    print(get_compilation_stage(QJIT, "QuantumCompilationPass"))
+    shutil.rmtree(QJIT.__name__)
+
+
+def get_custom_device_without(num_wires):
+    """Generate a custom device without gates in discards."""
+
```

Comment:
It doesn't look like this device is discarding any gates. Could we just use lightning.qubit or null.qubit instead then?

Reply:
Oh I see, it's for the device name. We should just inherit the CustomDevice from one of those two concrete devices then and only overwrite the name.

Reply:
Not instantiated, inherited. So something like:
```py
class CustomDevice(NullQubit):
    name = "oqd.cloud"
```

(Although you can definitely also modify class properties in Python, I just didn't suggest it because it would be a "global" change.)

Reply:
Updated :+1: